api/src/main/java/org/apache/iceberg/expressions/ExpressionUtil.java

```diff
@@ -502,19 +500,21 @@ public <T> String predicate(UnboundPredicate<T> pred) {
 
   private static List<String> abbreviateValues(List<String> sanitizedValues) {
     if (sanitizedValues.size() >= LONG_IN_PREDICATE_ABBREVIATION_THRESHOLD) {
-      Set<String> distinctValues = ImmutableSet.copyOf(sanitizedValues);
-      if (distinctValues.size()
-          <= sanitizedValues.size() - LONG_IN_PREDICATE_ABBREVIATION_MIN_GAIN) {
-        List<String> abbreviatedList = Lists.newArrayListWithCapacity(distinctValues.size() + 1);
-        abbreviatedList.addAll(distinctValues);
+      List<String> distinctValues = ImmutableSet.copyOf(sanitizedValues).asList();
+      List<String> abbreviatedList =
+          Lists.newArrayListWithCapacity(LONG_IN_PREDICATE_ABBREVIATION_THRESHOLD + 1);
+      int abbreviatedListSize =
+          Math.min(distinctValues.size(), LONG_IN_PREDICATE_ABBREVIATION_THRESHOLD);
```

Comment:
- abbreviatedListSize is not the size of the abbreviatedList list, so rename to abbreviationSize
- initialize the size first, so that the abbreviatedList can be sized as the int + 1
  - for example, if there are only 2 distinct values (hashes), the list should be inited with size 3, not 11

Reply:
done